### PR TITLE
[JENKINS-74858]  Added validation for Password length in FIPS mode

### PR DESCRIPTION
**Pull request by tejasdrolia**
See [JENKINS-74858](https://issues.jenkins.io/browse/JENKINS-74858).

Made changes to ensure that an exisiting user is now unable to alter the password to less than 14 characters in length when operating in FIPS mode.

### Testing done

**Automated Unit Testing**
Implemented unit tests to verify FIPS compliance for password changes:

1. Ensured that passwords shorter than 14 characters are rejected when FIPS mode is enabled.
2. Verified that mismatched passwords between user.password and user.password2 trigger appropriate errors.

**Manual Testing**
Did Manual testing to verify my changes to ensure that everything is working properly.

------------------------------------------------------------
**Without FIPS mode:** Shorter password is being accepted.
Screenshot 2024-11-20 at 12 37 33 PM

-------------------------------------------------------------
**With FIPS mode:** Shorter password is not being accepted.
Screenshot 2024-11-20 at 12 41 00 PM
Screenshot 2024-11-20 at 12 41 09 PM

### Proposed changelog entries

- Added Password validation to ensure that existing users cannot create a password of less than 14 characters in length(when in FIPS mode).

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java
Comment (PereBueno) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:843]: Maybe doing the check only in `user.password1` would make more sense? 
If lengths don't match the problem is that passwords are different and we would want the message in https://github.com/jenkinsci/jenkins/pull/9995/files#diff-18f9f27db9827aa6d004eb9000a9e27ec37197ae211d16346d0835f7c84fecd7R859 to be shown instead

File: test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java
Comment (daniel-beck) [inline test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java:34]: This looks like unnecessary use of mocks, prefer `WebClient`.

**Review by daniel-beck — COMMENTED** [review]
Would be nice if newly added tests covered the success case (long + matching password) in addition to not relying on mocks.

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java
Comment (tejasdrolia) [reply core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: Tried fixing in recent commit.

File: test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java
Comment (tejasdrolia) [reply test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java]: Fixed in recent commit

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java
Comment (fcojfernandez) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:847]: nit: merge the two conditions into a single one
```suggestion
                if (FIPS140.useCompliantAlgorithms() && pwd.length()< FIPS_PASSWORD_LENGTH) {
                    throw new FormException(Messages.HudsonPrivateSecurityRealm_CreateAccount_FIPS_PasswordLengthInvalid(), "user.password1");
                }

```

File: test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java
Comment (fcojfernandez) [inline test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java:28]: Do not use wildcards in the imports. Add as many imports as you need

File: test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java
Comment (fcojfernandez) [inline test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java:33]: Don't change the format of the file, please. Revert the  deletion. This together with the wildcard makes me think you should review your IDE settings

File: test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java
Comment (fcojfernandez) [inline test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java:160]: how is this possible? this password is invalid in FIPS mode (shorter than 14 characteres).

File: test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java
Comment (fcojfernandez) [inline test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java:175]: See the `userLoginAfterEnablingFIPS` method above in this test class. Probably you should check the log message as well. That method is doing

```
assertThat(getLogRecords(), hasItem(incorrectHashingLogEntry()));
```

**Review by fcojfernandez — COMMENTED** [review]
You have spotless issues, just execute `mvn clean install -DskipTests` to see and fix them

(haven't see them clearly in the CI logs, but they seem to be related to my comments about the empty line and the wildcard)

File: test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java
Comment (tejasdrolia) [reply test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java:179]: Apparently , there are no logs events during the test execution.

**Review by pankajy-dev — APPROVED** [review]
LGTM

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java
Comment (Priya-CB) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:842]: Is the code formatted? For instance, there's no space between pwd.length() and <

Comment (tejasdrolia) [reply core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:842]: Formatting is not an issue here I guess.
We have no spotless issue , so it's fine in my opinion.

File: test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java
Comment (tejasdrolia) [reply test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java:164]: https://cloudbees.atlassian.net/browse/BEE-53803

Created a ticket for this as discussed.
We also need to discuss about potential impact this ticket can have.

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java
Comment (fcojfernandez) [reply core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:842]: that is not spotless issue doesn't mean we should not keep the same formatting as in the rest of the file. Please fix the white space
```suggestion
                if (FIPS140.useCompliantAlgorithms() && pwd.length() < FIPS_PASSWORD_LENGTH) {
```

File: test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java
Comment (fcojfernandez) [reply test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java:179]: there must be as a FormException is thrown. If you mean that `getLogRecords()` is returning nothing you might be missing the `setupLogRecorder();`

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java
Comment (tejasdrolia) [reply core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: done

File: test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java
Comment (tejasdrolia) [reply test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java]: done

File: test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java
Comment (tejasdrolia) [reply test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java]: done

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java
Comment (tejasdrolia) [reply core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: done

**Review by fcojfernandez — APPROVED** [review]

**Review by olamy — APPROVED** [review]

**Comment by tejasdrolia** [comment]
@jenkinsci/core-pr-reviewers
Can you guys help in merging the PR?

File: test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java
Comment (jtnord) [reply test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java:164]: This is because FIPS compliant encryption for PKKDF is not in use.  This FIPS test merley sets the flag.  if you use a FIPS compiant implementation of PBKDF2 you should get an exception here.

@olamy added some code to the jenkins-test-harness to use FIPS compiant providers in tests for FIPS so if you swap to that it should blow up.  
it is still likely there is missing validation for the direct admin to create a user, but that should just be missing validation.

In any case - you should assume that you can not create this user with a short password here, create the user with an ok password and then try to change it to an invalid password.

**Review by jtnord — CHANGES_REQUESTED** [review]

File: test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java
Comment (jtnord) [inline test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java:191]: start with a valid password.

File: test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java
Comment (jtnord) [inline test/src/test/java/hudson/security/HudsonPrivateSecurityRealmFIPSTest.java:219]: start with a valid password for the SUT

**Review by jtnord — APPROVED** [review]

**Review by BorisYaoA — APPROVED** [review]

**Comment by fcojfernandez** [comment]
@jenkinsci/core-pr-reviewers any chance to get help to get this merged?

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java
Comment (res0nance) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:843]: Isn't the form field here `user.password` instead of `user.password1` as per line 834?

Comment (tejasdrolia) [reply core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java:843]: That's right , Thank you for pointing out!!

**Review by res0nance — APPROVED** [review]

**Comment by res0nance** [comment]
LGTM, I've set the ready for merge label. I'll be away for a few days but someone else can feel free to merge this in 24 hours if nothing comes up. Else I'll merge this when I'm back

**Pull request merged**
